plasmapy/diagnostics/tests/test_charged_particle_radiography.py

```diff
@@ -459,26 +458,203 @@ def test_run_options():
+        with numpy.inf when the intensity is zero.
+        """
+        bins = (200, 60)
+        size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
+
+        intensity_results = cpr.synthetic_radiograph(
+            self.sim_results, size=size, bins=bins
+        )
+        od_results = cpr.synthetic_radiograph(
+            self.sim_results, size=size, bins=bins, optical_density=True
+        )
+
+        assert np.allclose(intensity_results[0], od_results[0])
+        assert np.allclose(intensity_results[1], od_results[1])
+
+        intensity = intensity_results[2]
+        zero_mask = intensity == 0
+        i0 = np.mean(intensity[~zero_mask])
+        od = -np.log10(intensity / i0)
+
+        assert np.allclose(od[~zero_mask], od_results[2][~zero_mask])
+        assert np.all(np.isposinf(od_results[2][zero_mask]))
+
+
+def test_saving_output(tmp_path):
+
+    path = os.path.join(tmp_path, "temp.npz")
 
-    # CREATE A RADIOGRAPH OBJECT
     grid = _test_grid("electrostatic_gaussian_sphere", num=50)
     source = (0 * u.mm, -10 * u.mm, 0 * u.mm)
     detector = (0 * u.mm, 200 * u.mm, 0 * u.mm)
 
-    sim = prad.SyntheticProtonRadiograph(grid, source, detector, verbose=False)
+    sim = cpr.Tracker(grid, source, detector, verbose=False)
+
+    # Test that output cannot be saved prior to running
+    with pytest.raises(RuntimeError):
+        sim.results_dict
+
     sim.create_particles(1e4, 3 * u.MeV, max_theta=10 * u.deg)
     sim.run(field_weighting="nearest neighbor")
 
-    size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
-    bins = [200, 60]
+    res1 = sim.results_dict
+
+    # Save result
+    sim.save_results(path)
+
+    # Load result
+    res2 = dict(np.load(path, "r", allow_pickle=True))
 
-    # Test size is None, default bins
-    h, v, i = sim.synthetic_radiograph()
+    assert np.all(res1["x"] == res2["x"])
 
-    # Test optical density
-    h, v, i = sim.synthetic_radiograph(size=size, bins=bins, optical_density=True)
+
+def test_cannot_modify_simulation_after_running():
+
+    grid = _test_grid("electrostatic_gaussian_sphere", num=50)
+    source = (0 * u.mm, -10 * u.mm, 0 * u.mm)
+    detector = (0 * u.mm, 200 * u.mm, 0 * u.mm)
+    sim = cpr.Tracker(grid, source, detector, verbose=False)
+
+    # Test that changing the particles then requires the simulation to be
+    # run again prior to saving output
+    sim.create_particles(1e4, 3 * u.MeV, max_theta=10 * u.deg)
+    sim.run(field_weighting="nearest neighbor")
+
+    # Error from creating particles
+    with pytest.raises(RuntimeError):
+        sim.create_particles(1e4, 3 * u.MeV, max_theta=10 * u.deg)
+
+    # Error from loading particles
+    with pytest.raises(RuntimeError):
+        sim.load_particles(sim.x, sim.v)
+
+    # Error from adding wire mesh
+    with pytest.raises(RuntimeError):
+        sim.add_wire_mesh(
+            np.array([0, -2, 0]) * u.mm,
+            (2 * u.mm, 1.5 * u.mm),
+            9,
+            20 * u.um,
+        )
```

Comment:
```suggestion
@pytest.mark.parametrize(
    "case",
    ["creating particles", "loading particles", "adding a wire mesh"],
)
def test_cannot_modify_simulation_after_running(case):
    """
    Test that a Tracker objection can not be modified after it is
    run (Tracker.run).
    """

    sim = create_tracker_obj()
    sim.run(field_weighting="nearest neighbor")

    # Error from creating particles
    with pytest.raises(RuntimeError):
        if case == "creating particles":
            sim.create_particles(1e4, 3 * u.MeV, max_theta=10 * u.deg)
        elif case == "loading particles":
            sim.load_particles(sim.x, sim.v)
        elif case == "adding a wire mesh":
            sim.add_wire_mesh(
                np.array([0, -2, 0]) * u.mm,
                (2 * u.mm, 1.5 * u.mm),
                9,
                20 * u.um,
            )
        else:
            pytest.fail(f"Unrecognized test case '{case}'.")
```

`test_cannot_modify_simulation_after_running` needs a docstring and needs to be parametrize.